File: services/orders/feature_flags.py
"""OpenFeature (flagd) client for chaos / error-scenario gates."""
from __future__ import annotations

import logging
import os
from typing import Any

from openfeature import api
from openfeature.evaluation_context import EvaluationContext

logger = logging.getLogger("orders-service")

_FLAG_KEYS = [
    "chaos.bff_latency_ms",
    "chaos.catalog_latency_ms",
    "chaos.orders_latency_ms",
    "chaos.worker_latency_ms",
    "chaos.queue_lag_ms",
    "chaos.slow_close_ms",
    "chaos.fail_open_meteo",
    "chaos.fail_stripe",
    "chaos.fail_jsonplaceholder",
    "chaos.fail_catalog",
    "chaos.fail_publish",
]

_BOOL_FLAGS = {
    "chaos.fail_open_meteo",
    "chaos.fail_stripe",
    "chaos.fail_jsonplaceholder",
    "chaos.fail_catalog",
    "chaos.fail_publish",
}

_initialized = False


def _default_chaos() -> dict[str, Any]:
    return {
        key.replace("chaos.", "", 1): (False if key in _BOOL_FLAGS else 0)
        for key in _FLAG_KEYS
    }


def init_feature_flags() -> None:
    global _initialized
    if _initialized:
        return
    host = os.getenv("FLAGD_HOST", "flagd")
    port = int(os.getenv("FLAGD_PORT", "8013"))
    try:
        from openfeature.contrib.provider.flagd import FlagdProvider

        # openfeature-sdk >=0.8.2 passes flag_key=; provider must match (0.2.x; pin <=0.2.6 with OTel 1.29)
        api.set_provider(FlagdProvider(host=host, port=port, tls=False))
        _initialized = True
        logger.info("OpenFeature flagd provider ready at %s:%s", host, port)
    except Exception as exc:
        logger.warning("OpenFeature init failed (chaos flags disabled): %s", exc)


def _client():
    # openfeature-sdk uses domain= (not name=)
    return api.get_client(domain="orders-service")


def resolve_chaos(targeting_key: str | None = None) -> dict[str, Any]:
    """Evaluate chaos feature gates from flagd (outside control plane)."""
    out = _default_chaos()
    try:
        init_feature_flags()
        ctx = EvaluationContext(targeting_key=targeting_key or "anonymous")
        client = _client()
        for key in _FLAG_KEYS:
            short = key.replace("chaos.", "", 1)
            try:
                if key in _BOOL_FLAGS:
                    details = client.get_boolean_details(key, False, ctx)
                    if details.error_code:
                        logger.debug("flag %s error_code=%s", key, details.error_code)
                        continue
                    out[short] = bool(details.value)
                else:
                    details = client.get_integer_details(key, 0, ctx)
                    if details.error_code:
                        logger.debug("flag %s error_code=%s", key, details.error_code)
                        continue
                    out[short] = int(details.value or 0)
            except TypeError as exc:
                # Version skew between SDK and flagd provider — never break checkout
                logger.error(
                    "OpenFeature API mismatch evaluating %s (upgrade openfeature packages): %s",
                    key,
                    exc,
                )
                break
            except Exception as exc:
                logger.warning("flag eval failed for %s: %s", key, exc)
    except Exception as exc:
        logger.warning("resolve_chaos failed; using safe defaults: %s", exc)
    return out

File: services/bff/src/featureFlags.js
import { OpenFeature, ProviderEvents } from '@openfeature/server-sdk';
import { FlagdProvider } from '@openfeature/flagd-provider';
import fs from 'node:fs/promises';

const FLAG_FILE = process.env.FLAGD_FILE || '/etc/flagd/chaos.flagd.json';

const LATENCY_KEYS = [
  'chaos.bff_latency_ms',
  'chaos.catalog_latency_ms',
  'chaos.orders_latency_ms',
  'chaos.worker_latency_ms',
  'chaos.queue_lag_ms',
  'chaos.slow_close_ms',
];

const BOOL_KEYS = [
  'chaos.fail_open_meteo',
  'chaos.fail_stripe',
  'chaos.fail_jsonplaceholder',
  'chaos.fail_catalog',
  'chaos.fail_publish',
];

let readyPromise;

function shortName(key) {
  return key.replace(/^chaos\./, '');
}

function defaultsFromDoc(doc) {
  const out = {};
  for (const key of LATENCY_KEYS) {
    const flag = doc?.flags?.[key];
    const variant = flag?.defaultVariant || 'off';
    out[shortName(key)] = Number(flag?.variants?.[variant] ?? 0) || 0;
  }
  for (const key of BOOL_KEYS) {
    const flag = doc?.flags?.[key];
    const variant = flag?.defaultVariant || 'off';
    out[shortName(key)] = Boolean(flag?.variants?.[variant] ?? false);
  }
  return out;
}

/** File-backed fallback so Gates UI works even if flagd RPC is briefly down. */
async function resolveChaosFromFile() {
  try {
    const raw = await fs.readFile(FLAG_FILE, 'utf8');
    return defaultsFromDoc(JSON.parse(raw));
  } catch {
    return defaultsFromDoc({ flags: {} });
  }
}

export function initFeatureFlags() {
  if (!readyPromise) {
    readyPromise = (async () => {
      const host = process.env.FLAGD_HOST || 'flagd';
      const port = Number(process.env.FLAGD_PORT || 8013);
      try {
        // setContext is synchronous in @openfeature/server-sdk (does not return a Promise).
        OpenFeature.setContext({ targetingKey: 'bff-service' });

        OpenFeature.addHandler(ProviderEvents.Error, (event) => {
          console.warn('[openfeature] provider error (continuing with defaults):', event?.message || event);
        });
        OpenFeature.addHandler(ProviderEvents.Ready, () => {
          console.log(`[openfeature] flagd ready at ${host}:${port}`);
        });

        const maybePromise = OpenFeature.setProvider(
          new FlagdProvider({
            host,
            port,
            tls: false,
            maxEventStreamRetries: 50,
          })
        );
        if (maybePromise && typeof maybePromise.then === 'function') {
          await maybePromise;
        }
        console.log(`OpenFeature flagd provider registered at ${host}:${port}`);
      } catch (err) {
        console.warn('OpenFeature init warning (file fallback enabled):', err.message);
      }
    })();
  }
  return readyPromise;
}

export async function resolveChaos(targetingKey = 'anonymous') {
  await initFeatureFlags();
  // BFF owns the shared flag file — use it as source of truth for gateway decisions
  // (flagd still serves orders/worker; file hot-reload keeps them in sync).
  void targetingKey;
  return resolveChaosFromFile();
}

function variantForNumber(value) {
  if (!value || value <= 0) return 'off';
  if (value >= 1000) return 'high';
  return 'low';
}

function numberVariantValue(flagKey, variant) {
  const defaults = {
    'chaos.bff_latency_ms': { off: 0, low: 500, high: 2000 },
    'chaos.catalog_latency_ms': { off: 0, low: 500, high: 2000 },
    'chaos.orders_latency_ms': { off: 0, low: 500, high: 2000 },
    'chaos.worker_latency_ms': { off: 0, low: 500, high: 2000 },
    'chaos.queue_lag_ms': { off: 0, low: 1000, high: 5000 },
    'chaos.slow_close_ms': { off: 0, low: 1000, high: 3000 },
  };
  return defaults[flagKey]?.[variant] ?? 0;
}

/**
 * Update flagd file on shared volume. flagd watches the file and hot-reloads.
 */
export async function updateChaosFlags(chaos) {
  const raw = await fs.readFile(FLAG_FILE, 'utf8');
  const doc = JSON.parse(raw);
  if (!doc.flags) doc.flags = {};

  for (const key of LATENCY_KEYS) {
    const short = shortName(key);
    if (chaos[short] === undefined) continue;
    const ms = Number(chaos[short]) || 0;
    let variant = variantForNumber(ms);
    const variants = {
      off: 0,
      low: numberVariantValue(key, 'low'),
      high: numberVariantValue(key, 'high'),
    };
    if (ms > 0 && ms !== variants.low && ms !== variants.high) {
      variants.custom = ms;
      variant = 'custom';
    }
    doc.flags[key] = {
      state: 'ENABLED',
      variants,
      defaultVariant: variant,
    };
  }

  for (const key of BOOL_KEYS) {
    const short = shortName(key);
    if (chaos[short] === undefined) continue;
    const on = !!chaos[short];
    doc.flags[key] = {
      state: 'ENABLED',
      variants: { on: true, off: false },
      defaultVariant: on ? 'on' : 'off',
    };
  }

  const tmp = `${FLAG_FILE}.tmp`;
  await fs.writeFile(tmp, `${JSON.stringify(doc, null, 2)}\n`, 'utf8');
  await fs.rename(tmp, FLAG_FILE);
  await new Promise((r) => setTimeout(r, 400));
  return resolveChaos('flag-admin');
}

export async function readFlagFile() {
  const raw = await fs.readFile(FLAG_FILE, 'utf8');
  return JSON.parse(raw);
}

export { FLAG_FILE, LATENCY_KEYS, BOOL_KEYS };
